075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+3. On the **Nth** reconciliation (scaling up is done, rebalancing can start):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, but finally looking at the involved `StrimziPodSet`(s), it gets `readyPods` equals to `pods` so the scaling up is done. A new `KafkaRebalance` custom resource is created and the `Kafka.status.cruiseControl.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`). The `KafkaRebalanceAssemblyOperator` takes care of handling the rebalancing.
+4. From the **(N+1)th** to the **(M-1)th** reconciliation (rebalancing is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers are added this time. 
+     * the `autoRebalancing()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (we are assuming auto-approval). The rebalancing status is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+5. On the **Mth** reconciliation (rebalancing id done):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed).
+
+During the above flow, errors can happen after the scaling up:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+
+It's the user who should deal with this scenario:
+
+* Checking the reason why the `KafkaRebalance` is in the `NotReady` state and :
+  * deleting the custom resource if they are not interested in the rebalancing anymore. Scaling up is done already.
+  * setting the `strimzi.io/rebalance: refresh` annotation on the `KafkaRebalance` custom resource to try the rebalancing again.
+
+### Auto-rebalancing on scaling down
+
+The following flow is triggered when the user decreases the `replicas` field within a `Kafka` or `KafkaNodePool` custom resource in order to scale down the cluster by removing brokers.
+
+1. On the **1st** reconciliation (scaling down is skipped, rebalancing started):
+     * the check on scaling down the cluster fails, but it doesn't have to revert the `replicas` back because the `Kafka.spec.cruiseControl.autoRebalance` is provided. So the reconciliation doesn't fail.
```

Comment:
The sentence is correct, because today it's the cluster operator which reverts the change on replicas when the check on scaling down fails. It's not the user.

Reply:
Ah, I was not expecting this TBH. I expect a warning, like when a user tries to increase the KafkaTopic replicas without CC. In that case, the user is in charge to reverting the spec to get rid if the warning.

Reply:
This is how it's implemented today. You are blocked on doing the scale down and the operator reverts back otherwise the process will move forward.

Reply:
Ok, this makes sense, thanks.